autoprecompiles/src/lib.rs

```diff
@@ -756,142 +757,142 @@ pub fn optimize_exec_bus<T: FieldElement>(mut machine: SymbolicMachine<T>) -> Sy
     machine
 }
 
+#[derive(Default)]
+struct CompilationState<T> {
+    apc: SymbolicMachine<T>,
+    col_subs: Vec<Vec<u64>>,
+    global_idx: u64,
+}
+
 pub fn generate_precompile<T: FieldElement>(
     statements: &[SymbolicInstructionStatement<T>],
     instruction_kinds: &BTreeMap<String, InstructionKind>,
     instruction_machines: &BTreeMap<String, SymbolicMachine<T>>,
 ) -> (SymbolicMachine<T>, Vec<Vec<u64>>) {
-    let mut constraints: Vec<SymbolicConstraint<T>> = Vec::new();
-    let mut bus_interactions: Vec<SymbolicBusInteraction<T>> = Vec::new();
-    let mut col_subs: Vec<Vec<u64>> = Vec::new();
-    let mut global_idx: u64 = 3;
-
-    for (i, instr) in statements.iter().enumerate() {
-        match instruction_kinds.get(&instr.name).unwrap() {
-            InstructionKind::Normal
-            | InstructionKind::UnconditionalBranch
-            | InstructionKind::ConditionalBranch => {
-                let machine = instruction_machines.get(&instr.name).unwrap().clone();
-
-                let (next_global_idx, subs, machine) = powdr::reassign_ids(machine, global_idx, i);
-                global_idx = next_global_idx;
-
-                let pc_lookup: PcLookupBusInteraction<T> = machine
-                    .bus_interactions
-                    .iter()
-                    .filter_map(|bus_int| bus_int.clone().try_into().ok())
-                    .exactly_one()
-                    .expect("Expected single pc lookup");
-
-                let mut sub_map: BTreeMap<Column, AlgebraicExpression<T>> = Default::default();
-                let mut local_constraints: Vec<SymbolicConstraint<T>> = Vec::new();
-
-                let is_valid: AlgebraicExpression<T> = exec_receive(&machine).mult.clone();
-                let one = AlgebraicExpression::Number(1u64.into());
-                local_constraints.push((is_valid.clone() + one).into());
-
-                let mut sub_map_loadstore: BTreeMap<Column, AlgebraicExpression<T>> =
-                    Default::default();
-                if is_loadstore(instr.opcode) {
-                    sub_map_loadstore.extend(loadstore_chip_info(&machine, instr.opcode));
-                }
+    let CompilationState { apc, col_subs, .. } = statements.iter().enumerate().fold(
+        Default::default(),
+        |CompilationState {
+             mut apc,
+             mut col_subs,
+             mut global_idx,
+         },
+         (i, instr)| {
```

Comment:
I'm personally not a fan of huge blocks inside closures because they're so far apart and find it rather hard to read.

Reply:
imo extracting the inner block into a function would look better

Reply:
sure